src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+import {IJBPaymentTerminal} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBPaymentTerminal.sol";
+import {IJBRedemptionDelegate3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBRedemptionDelegate3_1_1.sol";
+import {IJBSingleTokenPaymentTerminal} from
+    "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBSingleTokenPaymentTerminal.sol";
+import {IJBSingleTokenPaymentTerminalStore3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBSingleTokenPaymentTerminalStore3_1_1.sol";
+import {JBDidPayData3_1_1} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBDidPayData3_1_1.sol";
+import {JBDidRedeemData3_1_1} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBDidRedeemData3_1_1.sol";
+import {JBFundingCycle} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBFundingCycle.sol";
+import {JBPayDelegateAllocation3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/structs/JBPayDelegateAllocation3_1_1.sol";
+import {JBPayParamsData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBPayParamsData.sol";
+import {JBRedeemParamsData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedeemParamsData.sol";
+import {JBRedemptionDelegateAllocation3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedemptionDelegateAllocation3_1_1.sol";
+import {JBTokenAmount} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBTokenAmount.sol";
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {UD60x18, ud, add, pow, powu, div, mul, wrap, unwrap} from "@prb/math/UD60x18.sol";
+
+/**
+ * @title Dominant Juice
+ * @author Armand Daigle
+ * @notice This contract is a JB Data Source, JB Pay Delegate, JB Redemption Delegate, and a Refund Bonus Escrow
+ * implementation. Campaigns, interchangeable with cycles from here on, that use this implementation as a Data Source
+ * will incorporate Alex Tabarrok's dominant assurance mechanism, which incentivize early pledgers with a campaign
+ * creator-provided refund bonus that can be withdrawn from this contract after a failed campaign.
+ * @dev This dominant assurance contract (DAC) uses the JB Data Source, Pay Delegate, and Pay Redemption interfaces and
+ * extends the functionality of a Juicebox project. Once a JB project is created, the constructor stores project data
+ * and the pertinent Juicebox architecture contracts. This contract is written for a single cycle campaign. If projects
+ * want dominant assurance on more cycles, refactoring is required.
+ */
+contract DominantJuice is
+    IJBFundingCycleDataSource3_1_1,
+    IJBPayDelegate3_1_1,
+    IJBRedemptionDelegate3_1_1,
+    AccessControl
+{
+    bytes32 private constant CAMPAIGN_MANAGER_ROLE = keccak256("CAMPAIGN_MANAGER_ROLE");
+
+    /// Juicebox Contracts
+    IJBController3_1 public immutable controller;
```

Comment:
do these have to be at the top level for Juicebox? Also, do they need to be public? if not, could make them private and inside a struct

Reply:
I can put them in a struct if you’d like. Will have to refactor a bunch in the tests I believe. See my response in the campaignInfo comment.